aea/protocols/base.py

```diff
@@ -251,21 +269,79 @@ class ProtobufSerializer(Serializer):
 
     @staticmethod
     def encode(msg: Message) -> bytes:
-        """Encode a message into bytes using Protobuf."""
-        body_json = Struct()
-        body_json.update(msg.body)  # pylint: disable=no-member
-        body_bytes = body_json.SerializeToString()
-        return body_bytes
+        """
+        Encode a message into bytes using Protobuf.
+
+        - if one of message_id, target and dialogue_reference are not defined,
+          serialize only the message body/
+        - otherwise, extract those fields from the body and instantiate
+          a Message struct.
+        """
+        message_pb = ProtobufMessage()
+        if msg.has_dialogue_info:
+            dialogue_message_pb = DialogueMessage()
+            dialogue_message_pb.message_id = msg.message_id
+            dialogue_message_pb.dialogue_starter_reference = msg.dialogue_reference[0]
+            dialogue_message_pb.dialogue_responder_reference = msg.dialogue_reference[1]
+            dialogue_message_pb.target = msg.target
+
+            new_body = copy(msg.body)
+            new_body.pop("message_id")
+            new_body.pop("dialogue_reference")
+            new_body.pop("target")
```

Comment:
we should never add the message_id, dialogue_ref and target to the body then?

Reply:
the problem is that they are in the body; they have been used above with `msg.message_id` etc., and we need to retrieve the rest of the body so we avoid to include the dialogue info fields twice

Reply:
That's correct